protocols/relay/src/priv_client/handler.rs

```diff
@@ -458,38 +351,167 @@ impl ConnectionHandler for Handler {
             return Poll::Ready(ConnectionHandlerEvent::Close(err));
         }
 
+        // Circuit connections
+        if let Poll::Ready(worker_res) = self.circuit_connection_futs.poll_unpin(cx) {
+            let res = match worker_res {
+                Ok(r) => r,
+                Err(Timeout { .. }) => {
+                    return Poll::Ready(ConnectionHandlerEvent::Close(StreamUpgradeError::Timeout));
+                }
+            };
+
+            let opt = match res {
+                Ok(Some(outbound_hop::Output::Circuit { limit })) => {
+                    Some(ConnectionHandlerEvent::NotifyBehaviour(
+                        Event::OutboundCircuitEstablished { limit },
+                    ))
+                }
+                Ok(None) => None,
+                Err(err) => {
+                    let res = match err {
+                        outbound_hop::UpgradeError::CircuitFailed(e) => {
+                            ConnectionHandlerEvent::NotifyBehaviour(
+                                Event::OutboundCircuitReqFailed {
+                                    error: StreamUpgradeError::Apply(e),
+                                },
+                            )
+                        }
+                        outbound_hop::UpgradeError::Fatal(e) => ConnectionHandlerEvent::Close(
+                            StreamUpgradeError::Apply(Either::Right(e)),
+                        ),
+                        outbound_hop::UpgradeError::ReservationFailed(_) => {
+                            unreachable!("do not emit `ReservationFailed` for connection")
```

Comment:
I am not very fond of having these `unreachable!` calls here. What do you think of just changing the API signature of `outbound_hop::handle_reserve_message_response` to return a `Result<Option>` by wrapping the return value in `Some`. Then we could merge these to `BoundedWorkers` and we wouldn't have these panics here.

WDYT?

Reply:
I think this is an excellent idea.